builtInFunctions/migrateUserName.go

```diff
@@ -0,0 +1,108 @@
+package builtInFunctions
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+)
+
+const userNameSep = "."
+
+type migrateUserName struct {
+	baseActiveHandler
+	gasCost         uint64
+	mapDnsAddresses map[string]struct{}
+	enableChange    bool
+	mutExecution    sync.RWMutex
+}
+
+// NewMigrateUserNameFunc returns a username built in function implementation
+func NewMigrateUserNameFunc(
+	gasCost uint64,
+	mapDnsAddresses map[string]struct{},
+	enableEpochsHandler vmcommon.EnableEpochsHandler,
+) (*migrateUserName, error) {
+	if mapDnsAddresses == nil {
+		return nil, ErrNilDnsAddresses
+	}
+	if check.IfNil(enableEpochsHandler) {
+		return nil, ErrNilEnableEpochsHandler
+	}
+
+	m := &migrateUserName{
+		gasCost:         gasCost,
+		mapDnsAddresses: make(map[string]struct{}, len(mapDnsAddresses)),
+	}
+	for key := range mapDnsAddresses {
+		m.mapDnsAddresses[key] = struct{}{}
+	}
+	m.baseActiveHandler.activeHandler = enableEpochsHandler.IsMigrateUsernameEnabled
+
+	return m, nil
+}
+
+// SetNewGasConfig is called whenever gas cost is changed
+func (m *migrateUserName) SetNewGasConfig(gasCost *vmcommon.GasCost) {
+	if gasCost == nil {
+		return
+	}
+
+	m.mutExecution.Lock()
+	m.gasCost = gasCost.BuiltInCost.SaveUserName
+	m.mutExecution.Unlock()
+}
+
+// ProcessBuiltinFunction sets the username to the account if it is allowed
+func (m *migrateUserName) ProcessBuiltinFunction(
```

Comment:
[question]

Just a question to make this clear for myself.
Although the caller can only be a DNS address (therefore trusted), should we add anyway checks for new username not to contain another `.`?

Reply:
Yes. There is an edge case there, where someone could steal your old username when migrating through the 256 DNS cotnracts.